File: expressjs/src/auth/permissions.ts
import type { UserContext } from "./auth.js";
import type { Registry } from "../metadata/registry.js";
import type { Permission, PermissionCondition } from "../metadata/permission.js";
import type { WhereClause } from "../engine/query.js";
import { AppError } from "../engine/errors.js";

export function checkPermission(
  user: UserContext | undefined,
  entity: string,
  action: string,
  registry: Registry,
  currentRecord: Record<string, any> | null,
): void {
  if (!user) {
    throw new AppError("UNAUTHORIZED", 401, "Authentication required");
  }

  // Admin bypasses all permission checks
  if (user.roles.includes("admin")) {
    return;
  }

  const policies = registry.getPermissions(entity, action);
  if (policies.length === 0) {
    throw new AppError(
      "FORBIDDEN",
      403,
      `No permission for ${action} on ${entity}`,
    );
  }

  // Check each policy — if ANY passes, the action is allowed
  for (const p of policies) {
    if (!hasRoleIntersection(user.roles, p.roles)) {
      continue;
    }
    // Role matches — now check conditions
    if (!p.conditions || p.conditions.length === 0) {
      return; // No conditions, role match is sufficient
    }
    if (currentRecord && evaluateConditions(p.conditions, currentRecord)) {
      return;
    }
    // For create, there's no current record — conditions don't apply
    if (!currentRecord && (action === "create" || action === "read")) {
      return;
    }
  }

  throw new AppError(
    "FORBIDDEN",
    403,
    `Permission denied for ${action} on ${entity}`,
  );
}

export function getReadFilters(
  user: UserContext | undefined,
  entity: string,
  registry: Registry,
): WhereClause[] {
  if (!user || user.roles.includes("admin")) {
    return [];
  }

  const policies = registry.getPermissions(entity, "read");
  if (policies.length === 0) {
    return [];
  }

  const filters: WhereClause[] = [];
  for (const p of policies) {
    if (!hasRoleIntersection(user.roles, p.roles)) {
      continue;
    }
    if (p.conditions) {
      for (const cond of p.conditions) {
        filters.push({
          field: cond.field,
          operator: cond.operator,
          value: cond.value,
        });
      }
    }
  }
  return filters;
}

function hasRoleIntersection(userRoles: string[], policyRoles: string[]): boolean {
  for (const ur of userRoles) {
    for (const pr of policyRoles) {
      if (ur.toLowerCase() === pr.toLowerCase()) {
        return true;
      }
    }
  }
  return false;
}

function evaluateConditions(
  conditions: PermissionCondition[],
  record: Record<string, any>,
): boolean {
  for (const cond of conditions) {
    const val = record[cond.field];
    if (val === undefined) return false;
    if (!evaluateCondition(cond.operator, val, cond.value)) {
      return false;
    }
  }
  return true;
}

function evaluateCondition(operator: string, recordVal: any, condVal: any): boolean {
  switch (operator) {
    case "eq":
      return String(recordVal) === String(condVal);
    case "neq":
      return String(recordVal) !== String(condVal);
    case "in":
      return valueInList(recordVal, condVal);
    case "not_in":
      return !valueInList(recordVal, condVal);
    case "gt":
      return compareNumeric(recordVal, condVal) > 0;
    case "gte":
      return compareNumeric(recordVal, condVal) >= 0;
    case "lt":
      return compareNumeric(recordVal, condVal) < 0;
    case "lte":
      return compareNumeric(recordVal, condVal) <= 0;
    default:
      return false;
  }
}

function valueInList(val: any, list: any): boolean {
  if (!Array.isArray(list)) return false;
  const valStr = String(val);
  return list.some((item) => String(item) === valStr);
}

function compareNumeric(a: any, b: any): number {
  const fa = Number(a) || 0;
  const fb = Number(b) || 0;
  return fa - fb;
}

File: golang/internal/engine/permissions.go
package engine

import (
	"fmt"
	"strings"

	"rocket-backend/internal/metadata"
)

// CheckPermission verifies that the user is allowed to perform the given action
// on the given entity. For update/delete, currentRecord is the existing record
// to check conditions against. Returns nil if allowed, or a FORBIDDEN AppError.
func CheckPermission(user *metadata.UserContext, entity, action string, reg *metadata.Registry, currentRecord map[string]any) error {
	if user == nil {
		return UnauthorizedError("Authentication required")
	}

	// Admin bypasses all permission checks
	if user.IsAdmin() {
		return nil
	}

	policies := reg.GetPermissions(entity, action)
	if len(policies) == 0 {
		return ForbiddenError(fmt.Sprintf("No permission for %s on %s", action, entity))
	}

	// Check each policy — if ANY passes, the action is allowed
	for _, p := range policies {
		if !hasRoleIntersection(user.Roles, p.Roles) {
			continue
		}
		// Role matches — now check conditions
		if len(p.Conditions) == 0 {
			return nil // No conditions, role match is sufficient
		}
		if currentRecord != nil && evaluateConditions(p.Conditions, currentRecord) {
			return nil
		}
		// For create, there's no current record — conditions don't apply
		if currentRecord == nil && (action == "create" || action == "read") {
			return nil
		}
	}

	return ForbiddenError(fmt.Sprintf("Permission denied for %s on %s", action, entity))
}

// GetReadFilters returns extra WhereClause entries to inject into read queries
// for row-level security. Admin users get no filters (full access).
func GetReadFilters(user *metadata.UserContext, entity string, reg *metadata.Registry) []WhereClause {
	if user == nil || user.IsAdmin() {
		return nil
	}

	policies := reg.GetPermissions(entity, "read")
	if len(policies) == 0 {
		return nil
	}

	var filters []WhereClause
	for _, p := range policies {
		if !hasRoleIntersection(user.Roles, p.Roles) {
			continue
		}
		for _, cond := range p.Conditions {
			filters = append(filters, WhereClause{
				Field:    cond.Field,
				Operator: cond.Operator,
				Value:    cond.Value,
			})
		}
	}
	return filters
}

func hasRoleIntersection(userRoles, policyRoles []string) bool {
	for _, ur := range userRoles {
		for _, pr := range policyRoles {
			if strings.EqualFold(ur, pr) {
				return true
			}
		}
	}
	return false
}

func evaluateConditions(conditions []metadata.PermissionCondition, record map[string]any) bool {
	for _, cond := range conditions {
		val, ok := record[cond.Field]
		if !ok {
			return false
		}
		if !evaluateCondition(cond.Operator, val, cond.Value) {
			return false
		}
	}
	return true
}

func evaluateCondition(operator string, recordVal, condVal any) bool {
	switch operator {
	case "eq":
		return fmt.Sprintf("%v", recordVal) == fmt.Sprintf("%v", condVal)
	case "neq":
		return fmt.Sprintf("%v", recordVal) != fmt.Sprintf("%v", condVal)
	case "in":
		return valueInList(recordVal, condVal)
	case "not_in":
		return !valueInList(recordVal, condVal)
	case "gt":
		return compareNumeric(recordVal, condVal) > 0
	case "gte":
		return compareNumeric(recordVal, condVal) >= 0
	case "lt":
		return compareNumeric(recordVal, condVal) < 0
	case "lte":
		return compareNumeric(recordVal, condVal) <= 0
	default:
		return false
	}
}

func valueInList(val, list any) bool {
	valStr := fmt.Sprintf("%v", val)
	switch l := list.(type) {
	case []any:
		for _, item := range l {
			if fmt.Sprintf("%v", item) == valStr {
				return true
			}
		}
	case []string:
		for _, item := range l {
			if item == valStr {
				return true
			}
		}
	}
	return false
}

func compareNumeric(a, b any) int {
	fa := toFloat(a)
	fb := toFloat(b)
	if fa < fb {
		return -1
	}
	if fa > fb {
		return 1
	}
	return 0
}

func toFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case float32:
		return float64(n)
	case int:
		return float64(n)
	case int64:
		return float64(n)
	case int32:
		return float64(n)
	default:
		var f float64
		fmt.Sscanf(fmt.Sprintf("%v", v), "%f", &f)
		return f
	}
}
